system/protocols/generic/factory.py

```diff
@@ -0,0 +1,225 @@
+# coding=utf-8
+import importlib
+
+from twisted.internet import reactor
+from twisted.internet.protocol import ClientFactory
+
+from system.protocols.generic.protocol import Protocol
+from system.logging.logger import getLogger
+
+__author__ = 'Gareth Coles'
+
+
+class BaseFactory(ClientFactory):
+    """
+    A Twisted factory, for producing protocols. Now with less fuckery!
+
+    Subclass this in your protocol's *factory.Factory* class, and use it to
+    describe how your protocol is built and how it connects. If you find this
+    confusing, you may want to read the other docstrings in this class, and
+    take a look at the factory for the *irc* protocol.
+    """
+
+    protocol = None
+    protocol_module = None
+
+    reconnection_attempts = 0
+    reconnection_config = None
+
+    shutting_down = False
+
+    def __init__(self, protocol_name, config, factory_manager):
+        self.name = protocol_name
+        self.config = config
+        self.factory_manager = factory_manager
+
+        self.logger = getLogger("F: {}".format(self.name))
+        self.reconnection_config = factory_manager.main_config["reconnections"]
+
+    # Custom Ultros functions - must be overridden
+
+    def connect(self):
+        """
+        Called to begin the connection - this is only called once.
+
+        In this function, you should read whatever you need from your config
+        file to work out how to connect, and then use the reactor to do so,
+        passing *self* in as the factory. For example:
+
+        >>> reactor.connectTCP(
+        ...     self.config["network"]["hostname"],
+        ...     self.config["network"]["port"],
+        ...     self,
+        ...     self.config["network"]["timeout"],
+        ...     bindAddress=("0.0.0.0", 0)
+        ... )
+        >>>
+
+        You **must** override this method, or your factory will not work.
+        """
+        raise NotImplementedError()
+
+    # Custom Ultros functions - Not necessary to override in most cases
+
+    def load_module(self):
+        """
+        Load the protocol module, reloading if it's already loaded.
+
+        If you use the standard Ultros naming scheme, then you won't need to
+        override this. For reference, this function expects your protocol to
+        reside in *system.protocols.[name].protocol*.
+        """
+
+        try:
+            if self.protocol_module is None:
+                self.logger.debug("First-time setup, not reloading")
+                self.protocol_module = importlib.import_module(
+                    "system.protocols.{}.protocol".format(
+                        self.config["main"]["protocol-type"]
```

Comment:
There's a lot of chained config lookups. Perhaps there should be a sanity check done at the start of the method to give nicer error messages.